Add luacheck

Panquesito7 opened the pull request:
- Add `.luacheckrc` file.
- Add GitHub workflow.
- Fix many (stylistic) errors and wrong indentations.
- Add build status badge on `README.md`.

sfan5 commented on .github/workflows/check-release.yml:
```diff
@@ -0,0 +1,11 @@
+    steps:
+      - uses: actions/checkout@master
+      - name: lint
+        uses: Roang-zero1/factorio-mod-luacheck@master
```
I'd rather not rely on some other person maintaining a thing.

On travis, installing luacheck is [a few lines of shell](https://github.com/minetest/minetest_game/blob/master/.travis.yml), can Github Actions do this too?

Panquesito7 replied:
Sure; fixed.

sfan5 commented on .luacheckrc:
```diff
@@ -0,0 +1,28 @@
+    "err", "state",
+    "filename", "def",
+    "file", "name",
+}
```
why are these ignored?

Panquesito7 replied:
Test locally and you'll see what happens.

sfan5 commented on .luacheckrc:
```diff
@@ -0,0 +1,28 @@
+
+globals = {
+	"minetest"
+}
```
couldn't this be in `read_globals`?

sfan5 commented on worldedit/manipulations.lua:
```diff
@@ -217,8 +217,6 @@ function worldedit.copy2(pos1, pos2, off, meta_backwards)
 
 	mh.finish(dst_manip)
-	src_data = nil
-	dst_data = nil
```
This was intentional.

Panquesito7 replied:
Checking `worldedit\manipulations.lua`              2 warnings

    worldedit\manipulations.lua:220:2: value assigned to variable 'src_data' is unused
    worldedit\manipulations.lua:221:2: value assigned to variable 'dst_data' is unused

This can also be fixed by adding them to the `ignore` list.

sfan5 commented on worldedit/primitives.lua:
```diff
@@ -299,7 +299,6 @@ function worldedit.spiral(pos, length, height, spacer, node_name)
 		for index = 1, segment_length do
 			-- Move along the direction of the segment
 			i = i + stride_axis * sign
-			local column = i
```
Incorrect.

Panquesito7 replied:
`column` was registered previously; this shouldn't be deleted.

sfan5 replied:
The value of `i` changes between the previous assignment and this one.

sfan5 commented on worldedit/primitives.lua:
```diff
@@ -314,7 +313,6 @@ function worldedit.spiral(pos, length, height, spacer, node_name)
 	sign = -sign
 	for index = 1, segment_length do
 		i = i + stride_axis * sign
-		local column = i
```
Incorrect.

Panquesito7 replied:
Same here as above.

sfan5 commented on worldedit_commands/init.lua:
```diff
@@ -45,14 +45,14 @@ local function chatcommand_handler(cmd_name, name, param)
 		local count = def.nodes_needed(name, unpack(parsed))
 		safe_region(name, count, function()
-			local success, msg = def.func(name, unpack(parsed))
+			local msg = def.func(name, unpack(parsed))
```
Incorrect.
Just disable the `unused_secondaries` warning if it doesn't report anything useful.

Panquesito7 replied:
Checking `worldedit_commands\init.lua`              4 warnings

    worldedit_commands\init.lua:48:10: unused variable 'success'
    worldedit_commands\init.lua:48:10: shadowing upvalue 'success' on line 39
    worldedit_commands\init.lua:55:9: unused variable 'success'
    worldedit_commands\init.lua:55:9: shadowing definition of variable 'success' on line 39

coil0 replied:
You have to replace `success` with `_`.

sfan5 commented on worldedit_commands/init.lua:
```diff
@@ -45,14 +45,14 @@ local function chatcommand_handler(cmd_name, name, param)
 	else
 		-- no "safe region" check
-		local success, msg = def.func(name, unpack(parsed))
+		local msg = def.func(name, unpack(parsed))
```
same here

Panquesito7 replied:
Same here as above.

sfan5 commented on worldedit_gui/init.lua:
```diff
@@ -88,7 +88,7 @@ if minetest.global_exists("unified_inventory") then -- unified inventory install
 			return true
 		elseif fields.worldedit_gui_exit then --return to original page
-			local player = minetest.get_player_by_name(name)
+			player = minetest.get_player_by_name(name)
```
This line can removed actually.

Panquesito7 commented:
I've fixed many warnings and removed the ignore list.
There's just one error that needs fixing:

```
Checking `worldedit_gui\init.lua`                   1 warning

    worldedit_gui\init.lua:157:26: shadowing upvalue argument 'state' on line 147
```

sfan5 commented:
Sorry about your effort, but I don't see this PR going anywhere when most of your suggested edits just break the code.
I do not have time to explain basic Lua programming to you.
